File: sdks/android/src/main/kotlin/io/natlangchain/rra/Clients.kt
// SPDX-License-Identifier: FSL-1.1-ALv2
// Copyright 2025 Kase Branham

package io.natlangchain.rra

/**
 * Client for marketplace API operations.
 */
class MarketplaceClient internal constructor(private val client: RRAClient) {

    /**
     * List agents in the marketplace.
     */
    suspend fun listAgents(
        page: Int = 1,
        pageSize: Int = 20
    ): AgentListResponse = client.get(
        path = "/api/marketplace/agents",
        queryParams = mapOf(
            "page" to page.toString(),
            "page_size" to pageSize.toString()
        )
    )

    /**
     * Search for agents.
     */
    suspend fun search(params: MarketplaceSearchParams): AgentListResponse {
        val queryParams = mutableMapOf<String, String>()
        params.query?.let { queryParams["q"] = it }
        params.tags?.let { queryParams["tags"] = it.joinToString(",") }
        params.minPrice?.let { queryParams["min_price"] = it.toString() }
        params.maxPrice?.let { queryParams["max_price"] = it.toString() }
        params.sortBy?.let { queryParams["sort"] = it.name.lowercase() }
        params.page?.let { queryParams["page"] = it.toString() }
        params.pageSize?.let { queryParams["page_size"] = it.toString() }

        return client.get(path = "/api/marketplace/search", queryParams = queryParams)
    }

    /**
     * Get a specific agent by ID.
     */
    suspend fun getAgent(id: String): Agent = client.get(path = "/api/marketplace/agents/$id")

    /**
     * Get featured agents.
     */
    suspend fun getFeatured(): List<Agent> = client.get(path = "/api/marketplace/featured")

    /**
     * Get trending agents.
     */
    suspend fun getTrending(limit: Int = 10): List<Agent> = client.get(
        path = "/api/marketplace/trending",
        queryParams = mapOf("limit" to limit.toString())
    )
}

/**
 * Client for negotiation API operations.
 */
class NegotiationClient internal constructor(private val client: RRAClient) {

    /**
     * Start a new negotiation.
     */
    suspend fun start(request: StartNegotiationRequest): NegotiationSession =
        client.post(path = "/api/negotiate/start", body = request)

    /**
     * Send a message in a negotiation.
     */
    suspend fun sendMessage(request: SendMessageRequest): NegotiationMessage =
        client.post(path = "/api/negotiate/message", body = request)

    /**
     * Get negotiation session by ID.
     */
    suspend fun getSession(id: String): NegotiationSession =
        client.get(path = "/api/negotiate/session/$id")

    /**
     * Accept the current offer.
     */
    suspend fun acceptOffer(sessionId: String): NegotiationSession =
        client.post(
            path = "/api/negotiate/accept",
            body = mapOf("session_id" to sessionId)
        )

    /**
     * Reject the current offer.
     */
    suspend fun rejectOffer(sessionId: String, reason: String? = null): NegotiationSession {
        val body = mutableMapOf("session_id" to sessionId)
        reason?.let { body["reason"] = it }
        return client.post(path = "/api/negotiate/reject", body = body)
    }

    /**
     * Make a counter-offer.
     */
    suspend fun counterOffer(sessionId: String, offer: Offer): NegotiationSession =
        client.post(
            path = "/api/negotiate/counter",
            body = mapOf("session_id" to sessionId, "offer" to offer)
        )
}

/**
 * Client for streaming payment operations.
 */
class StreamingClient internal constructor(private val client: RRAClient) {

    /**
     * Create a new streaming license.
     */
    suspend fun createStream(request: CreateStreamRequest): StreamingLicense =
        client.post(path = "/api/streaming/create", body = request)

    /**
     * Activate a stream.
     */
    suspend fun activate(licenseId: String): StreamingLicense =
        client.post(path = "/api/streaming/activate/$licenseId", body = EmptyBody())

    /**
     * Stop a stream.
     */
    suspend fun stop(licenseId: String): StreamingLicense =
        client.post(path = "/api/streaming/stop/$licenseId", body = EmptyBody())

    /**
     * Get stream status.
     */
    suspend fun getStatus(licenseId: String): StreamingLicense =
        client.get(path = "/api/streaming/status/$licenseId")

    /**
     * Check access for a license.
     */
    suspend fun checkAccess(licenseId: String): Boolean {
        val response: AccessResponse = client.get(path = "/api/streaming/access/$licenseId")
        return response.accessGranted
    }

    /**
     * Get all licenses for a buyer.
     */
    suspend fun getLicenses(buyerAddress: String): List<StreamingLicense> = client.get(
        path = "/api/streaming/tokens",
        queryParams = mapOf("buyer" to buyerAddress)
    )
}

/**
 * Client for analytics operations.
 */
class AnalyticsClient internal constructor(private val client: RRAClient) {

    /**
     * Get analytics overview.
     */
    suspend fun getOverview(timeRange: TimeRange = TimeRange.WEEK): AnalyticsOverview = client.get(
        path = "/api/analytics/overview",
        queryParams = mapOf("time_range" to timeRange.value)
    )

    /**
     * Get analytics for a specific agent.
     */
    suspend fun getAgentAnalytics(
        agentId: String,
        timeRange: TimeRange = TimeRange.WEEK
    ): AnalyticsOverview = client.get(
        path = "/api/analytics/agent/$agentId",
        queryParams = mapOf("time_range" to timeRange.value)
    )

    /**
     * Record an analytics event.
     */
    suspend fun recordEvent(
        eventType: String,
        agentId: String,
        metadata: Map<String, String>? = null
    ) {
        @kotlinx.serialization.Serializable
        data class EventRequest(
            @kotlinx.serialization.SerialName("event_type") val eventType: String,
            @kotlinx.serialization.SerialName("agent_id") val agentId: String,
            val metadata: Map<String, String>? = null
        )
        client.post<EventRequest, Unit>(
            path = "/api/analytics/event",
            body = EventRequest(eventType, agentId, metadata)
        )
    }
}

/**
 * Client for widget operations.
 */
class WidgetClient internal constructor(private val client: RRAClient) {

    /**
     * Initialize a widget.
     */
    suspend fun initialize(request: WidgetInitRequest): WidgetConfig =
        client.post(path = "/api/widget/init", body = request)

    /**
     * Get widget configuration.
     */
    suspend fun getConfig(widgetId: String): WidgetConfig =
        client.get(path = "/api/widget/config/$widgetId")
}

File: sdks/ios/Sources/RRA/Clients.swift
// SPDX-License-Identifier: FSL-1.1-ALv2
// Copyright 2025 Kase Branham

import Foundation

// MARK: - Marketplace Client

/// Client for marketplace API operations.
public class MarketplaceClient {
    private weak var client: RRAClient?

    init(client: RRAClient) {
        self.client = client
    }

    /// List agents in the marketplace.
    public func listAgents(
        page: Int = 1,
        pageSize: Int = 20
    ) async throws -> AgentListResponse {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.get(
            path: "/api/marketplace/agents",
            queryItems: [
                URLQueryItem(name: "page", value: String(page)),
                URLQueryItem(name: "page_size", value: String(pageSize))
            ]
        )
    }

    /// Search for agents.
    public func search(_ params: MarketplaceSearchParams) async throws -> AgentListResponse {
        guard let client = client else { throw RRAError.invalidResponse }

        var queryItems: [URLQueryItem] = []
        if let query = params.query {
            queryItems.append(URLQueryItem(name: "q", value: query))
        }
        if let tags = params.tags {
            queryItems.append(URLQueryItem(name: "tags", value: tags.joined(separator: ",")))
        }
        if let minPrice = params.minPrice {
            queryItems.append(URLQueryItem(name: "min_price", value: String(minPrice)))
        }
        if let maxPrice = params.maxPrice {
            queryItems.append(URLQueryItem(name: "max_price", value: String(maxPrice)))
        }
        if let sortBy = params.sortBy {
            queryItems.append(URLQueryItem(name: "sort", value: sortBy.rawValue))
        }
        if let page = params.page {
            queryItems.append(URLQueryItem(name: "page", value: String(page)))
        }
        if let pageSize = params.pageSize {
            queryItems.append(URLQueryItem(name: "page_size", value: String(pageSize)))
        }

        return try await client.get(path: "/api/marketplace/search", queryItems: queryItems)
    }

    /// Get a specific agent by ID.
    public func getAgent(id: String) async throws -> Agent {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.get(path: "/api/marketplace/agents/\(id)")
    }

    /// Get featured agents.
    public func getFeatured() async throws -> [Agent] {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.get(path: "/api/marketplace/featured")
    }

    /// Get trending agents.
    public func getTrending(limit: Int = 10) async throws -> [Agent] {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.get(
            path: "/api/marketplace/trending",
            queryItems: [URLQueryItem(name: "limit", value: String(limit))]
        )
    }
}

// MARK: - Negotiation Client

/// Client for negotiation API operations.
public class NegotiationClient {
    private weak var client: RRAClient?

    init(client: RRAClient) {
        self.client = client
    }

    /// Start a new negotiation.
    public func start(_ request: StartNegotiationRequest) async throws -> NegotiationSession {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.post(path: "/api/negotiate/start", body: request)
    }

    /// Send a message in a negotiation.
    public func sendMessage(_ request: SendMessageRequest) async throws -> NegotiationMessage {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.post(path: "/api/negotiate/message", body: request)
    }

    /// Get negotiation session by ID.
    public func getSession(id: String) async throws -> NegotiationSession {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.get(path: "/api/negotiate/session/\(id)")
    }

    /// Accept the current offer.
    public func acceptOffer(sessionId: String) async throws -> NegotiationSession {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.post(
            path: "/api/negotiate/accept",
            body: ["session_id": sessionId]
        )
    }

    /// Reject the current offer.
    public func rejectOffer(sessionId: String, reason: String? = nil) async throws -> NegotiationSession {
        guard let client = client else { throw RRAError.invalidResponse }
        var body: [String: String] = ["session_id": sessionId]
        if let reason = reason {
            body["reason"] = reason
        }
        return try await client.post(path: "/api/negotiate/reject", body: body)
    }

    /// Make a counter-offer.
    public func counterOffer(sessionId: String, offer: Offer) async throws -> NegotiationSession {
        guard let client = client else { throw RRAError.invalidResponse }
        struct CounterOfferRequest: Encodable {
            let sessionId: String
            let offer: Offer
        }
        return try await client.post(
            path: "/api/negotiate/counter",
            body: CounterOfferRequest(sessionId: sessionId, offer: offer)
        )
    }
}

// MARK: - Streaming Client

/// Client for streaming payment operations.
public class StreamingClient {
    private weak var client: RRAClient?

    init(client: RRAClient) {
        self.client = client
    }

    /// Create a new streaming license.
    public func createStream(_ request: CreateStreamRequest) async throws -> StreamingLicense {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.post(path: "/api/streaming/create", body: request)
    }

    /// Activate a stream.
    public func activate(licenseId: String) async throws -> StreamingLicense {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.post(
            path: "/api/streaming/activate/\(licenseId)",
            body: EmptyBody()
        )
    }

    /// Stop a stream.
    public func stop(licenseId: String) async throws -> StreamingLicense {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.post(
            path: "/api/streaming/stop/\(licenseId)",
            body: EmptyBody()
        )
    }

    /// Get stream status.
    public func getStatus(licenseId: String) async throws -> StreamingLicense {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.get(path: "/api/streaming/status/\(licenseId)")
    }

    /// Check access for a license.
    public func checkAccess(licenseId: String) async throws -> Bool {
        guard let client = client else { throw RRAError.invalidResponse }
        struct AccessResponse: Decodable {
            let accessGranted: Bool
        }
        let response: AccessResponse = try await client.get(path: "/api/streaming/access/\(licenseId)")
        return response.accessGranted
    }

    /// Get all licenses for a buyer.
    public func getLicenses(buyerAddress: String) async throws -> [StreamingLicense] {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.get(
            path: "/api/streaming/tokens",
            queryItems: [URLQueryItem(name: "buyer", value: buyerAddress)]
        )
    }
}

// MARK: - Analytics Client

/// Client for analytics operations.
public class AnalyticsClient {
    private weak var client: RRAClient?

    init(client: RRAClient) {
        self.client = client
    }

    /// Get analytics overview.
    public func getOverview(timeRange: TimeRange = .week) async throws -> AnalyticsOverview {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.get(
            path: "/api/analytics/overview",
            queryItems: [URLQueryItem(name: "time_range", value: timeRange.rawValue)]
        )
    }

    /// Get analytics for a specific agent.
    public func getAgentAnalytics(
        agentId: String,
        timeRange: TimeRange = .week
    ) async throws -> AnalyticsOverview {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.get(
            path: "/api/analytics/agent/\(agentId)",
            queryItems: [URLQueryItem(name: "time_range", value: timeRange.rawValue)]
        )
    }

    /// Record an analytics event.
    public func recordEvent(
        eventType: String,
        agentId: String,
        metadata: [String: String]? = nil
    ) async throws {
        guard let client = client else { throw RRAError.invalidResponse }
        struct EventRequest: Encodable {
            let eventType: String
            let agentId: String
            let metadata: [String: String]?
        }
        let _: EmptyResponse = try await client.post(
            path: "/api/analytics/event",
            body: EventRequest(eventType: eventType, agentId: agentId, metadata: metadata)
        )
    }
}

// MARK: - Widget Client

/// Client for widget operations.
public class WidgetClient {
    private weak var client: RRAClient?

    init(client: RRAClient) {
        self.client = client
    }

    /// Initialize a widget.
    public func initialize(_ request: WidgetInitRequest) async throws -> WidgetConfig {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.post(path: "/api/widget/init", body: request)
    }

    /// Get widget configuration.
    public func getConfig(widgetId: String) async throws -> WidgetConfig {
        guard let client = client else { throw RRAError.invalidResponse }
        return try await client.get(path: "/api/widget/config/\(widgetId)")
    }
}

// MARK: - Helper Types

struct EmptyBody: Encodable {}
struct EmptyResponse: Decodable {}
